### DISCUSSION TITLE
How to use SimpleIdServer as a headless service?

### DISCUSSION THREAD

**Discussion by AkiraVoid** [simpleidserver/SimpleIdServer#673] [Q&A]
I want to use SimpleIdServer to manage identities and control the access by only using its APIs and building a custom UI. But the docs haven't explained how to use its APIs (no API references). I currently don't know what each API is for, and what is needed to call that API (such as scopes, permissions, etc.).

**Comment by Alik2015** [comment]
See [this](https://github.com/simpleidserver/SimpleIdServer/discussions/610) it should help.

**Reply by AkiraVoid** [reply]
Thanks, it helps. But we are seeking a clearer and deeper explanation rather than just the endpoints about user registration since we need to rebuild the whole console UI and the sign-in/up wizard.

**Reply by AkiraVoid** [reply]
I am also planning to make a JavaScript library to help people leverage SimpleIdServer more easily in their end products.

**Reply by Alik2015** [reply]
If you look at all the endpoints you can figure out how to create your own ui on top.  We are also writing custom sigbin and register through our own app leveraging the server only for managing auth.

**Reply by simpleidserver** [reply]
Hello, first of all, thank you for your interest in SimpleIdServer :)
I would be very glad if you could participate in the development of a JavaScript library to help people develop their own UI client!

You can download the Postman project below; it contains the HTTP requests needed to execute the most important workflows, such as:
* Registering one user and their credentials.
* Registering one group.
* Registering one client.
* Assigning one client role to one group.
* Assigning one group to a user.

I hope this project can be useful and give you a better understanding of the different APIs :)

[SimpleIdServerManagement.postman_collection.json](https://github.com/simpleidserver/SimpleIdServer/files/13820769/SimpleIdServerManagement.postman_collection.json)

**Reply by Alik2015** [reply]
In order for the above postman to work You will need to retrieve the token as a request and not via the Authorization tab in postman as this will not work.  In the request you will need to send 

![image](https://github.com/simpleidserver/SimpleIdServer/assets/31630746/5378125b-fafd-4ec8-8f56-cc87885a6753)

@simpleidserver can you confirm this is the right way.  As just clicking the Authorization tab in postman will not work.

Also what is the best way to manage this token when doing lots of registrations?  Is it best to keep a cached version and check expiry or just get a new token every time you wish to perform an operation?

**Reply by simpleidserver** [reply]
Indeed, clicking on the Authorization tab is not enough; you must follow the actions below:

* Select the first `Get access token` request.
* Click on the `Authorization` tab and click on the `Get new access token` button.

![image](https://github.com/simpleidserver/SimpleIdServer/assets/10213388/01fd8e25-7b37-46b9-9d58-a345e900a6d1)

* Once the access token is received, select the request you are interested in executing, for example, `Add one user`.
* Choose the `Authorization tab` and enter the new token

![image](https://github.com/simpleidserver/SimpleIdServer/assets/10213388/8d2ec591-eae5-4826-b806-60ec093be117)

Postman retains the token in its cache, and an error message is displayed when the access token expires.

You can click on the `Refresh` link to renew the access token :).

**Comment by Alik2015** [comment]
How would this request look raw? For example using httpclient in dot net? And regarding the question about token management it best to cache  the token or get a new one every time.On 16 Jan 2024 15:37, SimpleIdServer ***@***.***> wrote:
Indeed, clicking on the Authorization tab is not enough; you must follow the actions below:
Select the first Get access token request.Click on the Authorization tab and click on the Get new access token button.
image.png (view on web)
Once the access token is received, select the request you are interested in executing, for example, Add one user.Choose the Authorization tab and enter the new token
image.png (view on web)
Postman retains the token in its cache, and an error message is displayed when the access token expires.
You can click on the Refresh link to renew the access token :).

**Reply by simpleidserver** [reply]
I have updated the Postman project and written the HTTP request clearly to obtain a valid access token, which will be utilized for interacting with different APIs :).
Additionally, the `client_credentials` grant_type has been enabled for the `postman` client.

[SimpleIdServerManagement.postman_collection.json](https://github.com/simpleidserver/SimpleIdServer/files/13954782/SimpleIdServerManagement.postman_collection.json)

Regarding the token, it is advisable to cache it and use the same until it expires. 
You can find a code sample here: https://github.com/simpleidserver/SimpleIdServer/blob/f384689d4c35e4bdefa83dcde76a6b330769f254/src/IdServer/SimpleIdServer.IdServer.Website/WebsiteHttpClientFactory.cs#L53

**Comment by thabart** [comment]
@Alik2015 

Your HTTP request is correct :)
Could you please check the following points?


  *   Ensure that the apitesting client is configured with the password grant type.

  *   Verify that the password for the specified username is active.

________________________________
De : Am ***@***.***>
Envoyé : mardi 30 janvier 2024 14:45
À : simpleidserver/SimpleIdServer ***@***.***>
Cc : Subscribed ***@***.***>
Objet : Re: [simpleidserver/SimpleIdServer] How to use SimpleIdServer as a headless service? (Discussion #673)


Assuming I have a user set up called testuser and client called apitesting for testing the api via unit tests set up as Machine client. I am trying to obtain a token for testuser using the postman but get an error. Might be missing a setting maybe?

image.png (view on web)<https://github.com/simpleidserver/SimpleIdServer/assets/31630746/aaa301e3-b3da-47e9-82c6-1f44f9ccfa0e>
